### ISSUE TITLE
Internal floating-point rounding error?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2636]
Migrated from [rt.perl.org#4326](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4326) (status was 'resolved')

Searchable as RT4326$

**Comment by p5pRT** [comment]
### From phil@inchip.com

#### Created by Philip.Andrew@stanfordalumni.org



I get strange results running the following Perl stub&#8203;:

&nbsp;    printf "%04d\\n"\, $ARGV\[0\]/$ARGV\[1\];

Results&#8203;:

&nbsp;    \# Command Run    =>    Output
&nbsp;    \# \-\-\-\-\-\-\-\-\-\-\-        \-\-\-\-\-\-
&nbsp;    stub\.pl 0\.55 0\.01   =>      0055
&nbsp;    stub\.pl 0\.56 0\.01   =>      0056
&nbsp;    stub\.pl 0\.57 0\.01   =>      0056    \<== HUH?
&nbsp;    stub\.pl 0\.58 0\.01   =>      0057    \<== HUH?
&nbsp;    stub\.pl 0\.59 0\.01   =>      0058    \<== HUH?
&nbsp;    stub\.pl 0\.60 0\.01   =>      0060
&nbsp;    stub\.pl 0\.61 0\.01   =>      0061
&nbsp;    stub\.pl 0\.055 0\.001 =>      0055
&nbsp;    stub\.pl 0\.056 0\.001 =>      0056
&nbsp;    stub\.pl 0\.057 0\.001 =>      0057
&nbsp;    stub\.pl 0\.058 0\.001 =>      0058
&nbsp;    stub\.pl 0\.059 0\.001 =>      0058    \<== HUH?
&nbsp;    stub\.pl 0\.060 0\.001 =>      0060
&nbsp;    stub\.pl 0\.061 0\.001 =>      0061

I can avoid this round\-off error by changing the 'printf' to

&nbsp;    printf "%04d\\n"\, \($ARGV\[0\]/$ARGV\[1\]\)\+\($ARGV\[1\]/10\);

\(adding 1/10 of my resolution to the number\)\, but what's the
point \-\- the round\-off is supposed to be correct\, isn't it?

I ran into this trying to create an encoding string from a real
number\, thereby eliminating the decimal point from the string
\(so that I could use the encoding in a file name\, and not have an
unwanted period added in the file name\)\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.0:

Configured by toolsadm at Thu Aug 17 18:36:22 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris
    uname='sunos chip 5.5.1 generic_103640-08 sun4u sparc sunw,ultra-1 '
    config_args='-de -A cc=gcc'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc=' gcc', optimize='-O', gccversion=2.95.2 19991024 (release)
    cppflags='-fno-strict-aliasing -I/usr/local/include -I/opt/gnu/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -I/opt/gnu/include '
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld=' gcc', ldflags =' -L/usr/local/lib -L/opt/gnu/lib '
    libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/usr/local/lib -L/opt/gnu/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/sun4-solaris
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/phil
    LANG (unset)
    LANGUAGE (unset)
    LC_COLLATE=en_US.ISO8859-1
    LC_CTYPE=en_US.ISO8859-1
    LC_MESSAGES=C
    LC_MONETARY=en_US.ISO8859-1
    LC_NUMERIC=en_US.ISO8859-1
    LC_TIME=en_US.ISO8859-1
    LD_LIBRARY_PATH=/usr/openwin/lib:/usr/local/lib:/opt/gnu/lib:/usr/lib
    LOGDIR (unset)
    PATH=.:/home/phil/bin:/home/phil/bin/solaris:/usr/local/bin:/opt/gnu/bin:/usr/bin:/usr/sbin:/usr/ccs/bin:/usr/ucb:/etc:/usr/dt/bin:/usr/openwin/bin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>This is a bug report for perl from Philip\.Andrew@&#8203;stanfordalumni\.org\,
>generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.

>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]

>I get strange results running the following Perl stub&#8203;:

>    printf "%04d\\n"\, $ARGV\[0\]/$ARGV\[1\];

>Results&#8203;:

>    \# Command Run    =>    Output
>    \# \-\-\-\-\-\-\-\-\-\-\-        \-\-\-\-\-\-
>    stub\.pl 0\.55 0\.01   =>      0055
>    stub\.pl 0\.56 0\.01   =>      0056
>    stub\.pl 0\.57 0\.01   =>      0056    \<== HUH?
>    stub\.pl 0\.58 0\.01   =>      0057    \<== HUH?
>    stub\.pl 0\.59 0\.01   =>      0058    \<== HUH?
>    stub\.pl 0\.60 0\.01   =>      0060
>    stub\.pl 0\.61 0\.01   =>      0061
>    stub\.pl 0\.055 0\.001 =>      0055
>    stub\.pl 0\.056 0\.001 =>      0056
>    stub\.pl 0\.057 0\.001 =>      0057
>    stub\.pl 0\.058 0\.001 =>      0058
>    stub\.pl 0\.059 0\.001 =>      0058    \<== HUH?
>    stub\.pl 0\.060 0\.001 =>      0060
>    stub\.pl 0\.061 0\.001 =>      0061

>I can avoid this round\-off error by changing the 'printf' to

>    printf "%04d\\n"\, \($ARGV\[0\]/$ARGV\[1\]\)\+\($ARGV\[1\]/10\);

>\(adding 1/10 of my resolution to the number\)\, but what's the
>point \-\- the round\-off is supposed to be correct\, isn't it?

Define "correct" and define "round" \-\- both within the rules and
regulations of binary floating\-point arithmetic\.

I'm afraid Perl has no bug\.  It just doesn't do what you want
it to do\.  

There is no exact representation of 0\.56&#8203;:

&nbsp;      printf "%\.20f\\n"\, \.57
&nbsp;    0\.56999999999999995115

Now watch&#8203;:

&nbsp;      printf "%\.20f\\n"\, \.57 \* 100
&nbsp;    56\.99999999999999289457

What would the int\(\)eger truncation be there?  Find the decimal
place\, and discard\.  Thus&#8203;:

&nbsp;      printf "%d\\n"\, \.57 \* 100
&nbsp;    56

&nbsp;      print int\(\.57 \* 100\)
&nbsp;    56

Now\, why does&#8203;:

&nbsp;      print \(\.57 \* 100\)
&nbsp;    57

Because print's default output format for numbers is such that 
you get some sort of rounding\.  

Of course\, this is still not always what you want \(IEEE's required
alternating "rounding" behavior on the halves\, for example\)\.

\-\-tom

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Philip Andrew \<Philip\.Andrew@&#8203;stanfordalumni\.org> writes&#8203;:
>
>\(adding 1/10 of my resolution to the number\)\, but what's the
>point \-\- the round\-off is supposed to be correct\, isn't it?

perl \(like the C it is built upon\) has float \-> int as truncate toward
zero not round\. Your '%04d' forces integer\. If you use %f or %g 
then round should be to nearest \(decimal\) and hence "correct"\.

**Comment by p5pRT** [comment]
### From @JohnPeacock

Depending on your actual needs\, you may want to consider not using FP
numbers at all\, due to the rouding problems you have noticed\.  You can
user Math&#8203;::BigFloat or \(my own\) Math&#8203;::FixedPrecision to do exact FP math
instead\.  Since both modules ultimately depend on Math&#8203;::Integer\, they
are currently not fast\, but you will get what you expect\.  M&#8203;:FP auto
rounds to a specified number of decimal places\, whereas M&#8203;::BF does not\.

Note that the Math&#8203;::BigFloat currently distributed with 5\.6\.0 has bugs;
a patched version should be in 5\.6\.1 \(hopefully RSN\)\.  It is also
included in Math&#8203;::FixedPrecision v0\.10\.

John Peacock

Philip Andrew wrote&#8203;:
> 
> This is a bug report for perl from Philip\.Andrew@&#8203;stanfordalumni\.org\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> I get strange results running the following Perl stub&#8203;:
> 
>     printf "%04d\\n"\, $ARGV\[0\]/$ARGV\[1\];
> 
> Results&#8203;:
> 
>     \# Command Run       =>      Output
>     \# \-\-\-\-\-\-\-\-\-\-\-               \-\-\-\-\-\-
>     stub\.pl 0\.55 0\.01   =>      0055
>     stub\.pl 0\.56 0\.01   =>      0056
>     stub\.pl 0\.57 0\.01   =>      0056    \<== HUH?
>     stub\.pl 0\.58 0\.01   =>      0057    \<== HUH?
>     stub\.pl 0\.59 0\.01   =>      0058    \<== HUH?
>     stub\.pl 0\.60 0\.01   =>      0060
>     stub\.pl 0\.61 0\.01   =>      0061
>     stub\.pl 0\.055 0\.001 =>      0055
>     stub\.pl 0\.056 0\.001 =>      0056
>     stub\.pl 0\.057 0\.001 =>      0057
>     stub\.pl 0\.058 0\.001 =>      0058
>     stub\.pl 0\.059 0\.001 =>      0058    \<== HUH?
>     stub\.pl 0\.060 0\.001 =>      0060
>     stub\.pl 0\.061 0\.001 =>      0061
> 
> I can avoid this round\-off error by changing the 'printf' to
> 
>     printf "%04d\\n"\, \($ARGV\[0\]/$ARGV\[1\]\)\+\($ARGV\[1\]/10\);
> 
> \(adding 1/10 of my resolution to the number\)\, but what's the
> point \-\- the round\-off is supposed to be correct\, isn't it?
> 
> I ran into this trying to create an encoding string from a real
> number\, thereby eliminating the decimal point from the string
> \(so that I could use the encoding in a file name\, and not have an
> unwanted period added in the file name\)\.
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl v5\.6\.0&#8203;:
> 
> Configured by toolsadm at Thu Aug 17 18&#8203;:36&#8203;:22 PDT 2000\.
> 
> Summary of my perl5 \(revision 5\.0 version 6 subversion 0\) configuration&#8203;:
>   Platform&#8203;:
>     osname=solaris\, osvers=2\.5\.1\, archname=sun4\-solaris
>     uname='sunos chip 5\.5\.1 generic\_103640\-08 sun4u sparc sunw\,ultra\-1 '
>     config\_args='\-de \-A cc=gcc'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>     useperlio=undef d\_sfio=undef uselargefiles=define
>     use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
>   Compiler&#8203;:
>     cc=' gcc'\, optimize='\-O'\, gccversion=2\.95\.2 19991024 \(release\)
>     cppflags='\-fno\-strict\-aliasing \-I/usr/local/include \-I/opt/gnu/include'
>     ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-I/opt/gnu/include '
>     stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
>     alignbytes=8\, usemymalloc=y\, prototype=define
>   Linker and Libraries&#8203;:
>     ld=' gcc'\, ldflags =' \-L/usr/local/lib \-L/opt/gnu/lib '
>     libpth=/usr/local/lib /opt/gnu/lib /lib /usr/lib /usr/ccs/lib
>     libs=\-lsocket \-lnsl \-ldl \-lm \-lc \-lcrypt \-lsec
>     libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib \-L/opt/gnu/lib'
> 
> Locally applied patches&#8203;:
> 
> 
> \-\-\-
> @&#8203;INC for perl v5\.6\.0&#8203;:
>     /usr/local/lib/perl5/5\.6\.0/sun4\-solaris
>     /usr/local/lib/perl5/5\.6\.0
>     /usr/local/lib/perl5/site\_perl/5\.6\.0/sun4\-solaris
>     /usr/local/lib/perl5/site\_perl/5\.6\.0
>     /usr/local/lib/perl5/site\_perl
>     \.
> 
> \-\-\-
> Environment for perl v5\.6\.0&#8203;:
>     HOME=/home/phil
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LC\_COLLATE=en\_US\.ISO8859\-1
>     LC\_CTYPE=en\_US\.ISO8859\-1
>     LC\_MESSAGES=C
>     LC\_MONETARY=en\_US\.ISO8859\-1
>     LC\_NUMERIC=en\_US\.ISO8859\-1
>     LC\_TIME=en\_US\.ISO8859\-1
>     LD\_LIBRARY\_PATH=/usr/openwin/lib&#8203;:/usr/local/lib&#8203;:/opt/gnu/lib&#8203;:/usr/lib
>     LOGDIR \(unset\)
>     PATH=\.&#8203;:/home/phil/bin&#8203;:/home/phil/bin/solaris&#8203;:/usr/local/bin&#8203;:/opt/gnu/bin&#8203;:/usr/bin&#8203;:/usr/sbin&#8203;:/usr/ccs/bin&#8203;:/usr/ucb&#8203;:/etc&#8203;:/usr/dt/bin&#8203;:/usr/openwin/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/local/bin/tcsh

**Comment by p5pRT** [comment]
### From The RT System itself

not a bug\, see comments by tom\.
